src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,251 @@
+#include "distributed/coordinator_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_logical_replication.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/pg_dist_rebalance_strategy.h"
+#include "distributed/pg_dist_shard.h"
+#include "distributed/reference_table_utils.h"
+#include "distributed/remote_commands.h"
+#include "distributed/resource_lock.h"
+#include "distributed/tuplestore.h"
+#include "distributed/utils/array_type.h"
+#include "distributed/worker_protocol.h"
+#include "nodes/pg_list.h"
+#include "postmaster/postmaster.h"
+
+
+PG_FUNCTION_INFO_V1(citus_auto_shard_split_start); 
+#define MAX_SHARD_SIZE 20000
+
+typedef struct ShardInfoData
+{
+    int64 shardsize;
+    int64 shardminvalue;
+    int64 shardmaxvalue;
+    int64 shardid ;
+    int64 nodeid ;
+    char *tablename;
+    char *distributionColumn;
+    char *datatype;
+
+}ShardInfoData;
+typedef ShardInfoData *shardInfo;
+
+char*
+ExecuteDatatypeQuery(const char *tablename, const char *distributionColumn)
+{
+    StringInfo DatatypeQuery = makeStringInfo();
+    appendStringInfo(DatatypeQuery,"SELECT pg_typeof(%s) AS data_type FROM %s LIMIT 1",distributionColumn,tablename);
+    ereport(LOG,errmsg("%s",DatatypeQuery->data));
+    if(SPI_connect() != SPI_OK_CONNECT)
+    {
+        elog(ERROR,"SPI_connect to the query failed");
+    }
+    if(SPI_exec(DatatypeQuery->data,0) != SPI_OK_SELECT)
+    {
+        elog(ERROR,"SPI_exec for the execution failed");
+    }
+    SPITupleTable *tupletable = SPI_tuptable;
+    HeapTuple tuple = tupletable->vals[0];
+
+    char *datatype = SPI_getvalue(tuple,tupletable->tupdesc,1);
+    char* result = pstrdup(datatype); 
+    SPI_freetuptable(tupletable);
+    SPI_finish();
+
+    return result;
+
+
+}
+
+StringInfo 
+GetSplitQuery(shardInfo shardinfo, char *averagehash)
+{
+    StringInfo hashsplitquery = makeStringInfo();
+    appendStringInfo(hashsplitquery,"SELECT citus_split_shard_by_split_points(%ld, ARRAY['%s'], ARRAY[%ld,%ld], 'block_writes')",
+        shardinfo->shardid, averagehash, shardinfo->nodeid, shardinfo->nodeid);      
+    return hashsplitquery;
+}
+
+
+
+char *
+ExecuteAvgHashQuery(shardInfo shardinfo )
+{
+
+    StringInfo AvgHashQuery = makeStringInfo();
+    appendStringInfo(AvgHashQuery, "SELECT avg(h)::int,count(*)"
+    " FROM (SELECT worker_hash(%s) h FROM %s tablesample system(10)"
+    " WHERE worker_hash(%s)>=%ld AND worker_hash(%s)<=%ld) s"
+    , shardinfo->distributionColumn, shardinfo->tablename , shardinfo->distributionColumn , shardinfo->shardminvalue , shardinfo->distributionColumn , shardinfo->shardmaxvalue
+    );
+    ereport(LOG,errmsg("%s",AvgHashQuery->data));
+    SPI_connect();
+    SPI_exec(AvgHashQuery->data,0);
+    SPITupleTable *tupletable = SPI_tuptable;
+    HeapTuple tuple = tupletable->vals[0];
+    char *average = SPI_getvalue(tuple,tupletable->tupdesc,1);
```

Comment:
Check for NULL value returned before using the pointer.

Reply:
fixed it